File: .claude/hooks/inject-deps-versions.mjs
#!/usr/bin/env node
// Injecte les versions réelles des dépendances dans chaque prompt (UserPromptSubmit)
import { readFileSync, existsSync } from 'fs';
import { fileURLToPath } from 'url';
import { join } from 'path';

const MAX_ENTRIES = 60; // borne le coût en tokens

function parsePackageJson(raw) {
  let pkg;
  try { pkg = JSON.parse(raw); } catch { return []; }
  const deps = { ...(pkg.dependencies || {}), ...(pkg.devDependencies || {}) };
  return Object.entries(deps).map(([name, version]) => `${name}@${version}`);
}

// Extraction best-effort des dépendances pyproject (PEP 621 [project] et Poetry).
function parsePyproject(raw) {
  const out = [];
  const block = raw.match(/dependencies\s*=\s*\[([\s\S]*?)\]/);
  if (block) {
    for (const m of block[1].matchAll(/["']([^"']+)["']/g)) out.push(m[1].trim());
  }
  return out;
}

export function run({ cwd = process.cwd(), readFile = readFileSync, fileExists = existsSync } = {}) {
  const entries = [];
  const pkgPath = join(cwd, 'package.json');
  if (fileExists(pkgPath)) {
    try { entries.push(...parsePackageJson(readFile(pkgPath, 'utf8'))); } catch {}
  }
  const pyPath = join(cwd, 'pyproject.toml');
  if (fileExists(pyPath)) {
    try { entries.push(...parsePyproject(readFile(pyPath, 'utf8'))); } catch {}
  }

  if (entries.length === 0) return null;

  const shown = entries.slice(0, MAX_ENTRIES);
  const more = entries.length > MAX_ENTRIES ? ` (+${entries.length - MAX_ENTRIES} more)` : '';
  return (
    '## Installed dependency versions\n' +
    'Use these exact versions — do not assume newer/older APIs:\n' +
    shown.map((e) => `- ${e}`).join('\n') + more + '\n'
  );
}

/* v8 ignore next 4 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const result = run();
  if (result) process.stdout.write(result);
}

File: .claude/hooks/inject-deps-versions.py
#!/usr/bin/env python3
# @hookstack user-prompt-inject-deps-versions
"""Injects real dependency versions into every prompt (UserPromptSubmit)."""
import json
import os
import re
import sys

MAX_ENTRIES = 60  # borne le coût en tokens


def _exists(path):
    return os.path.exists(path)


def _read_file(path):
    with open(path, "r", encoding="utf8") as f:
        return f.read()


def parse_package_json(raw):
    try:
        pkg = json.loads(raw)
    except Exception:
        return []
    deps = {}
    deps.update(pkg.get("dependencies") or {})
    deps.update(pkg.get("devDependencies") or {})
    return [f"{name}@{version}" for name, version in deps.items()]


# Extraction best-effort des dépendances pyproject (PEP 621 [project] et Poetry).
def parse_pyproject(raw):
    out = []
    match = re.search(r"dependencies\s*=\s*\[([\s\S]*?)\]", raw)
    if match:
        out.extend(m.strip() for m in re.findall(r"[\"']([^\"']+)[\"']", match.group(1)))
    return out


def run(*, cwd=None, read_file=None, file_exists=None):
    if cwd is None:
        cwd = os.getcwd()
    if read_file is None:
        read_file = _read_file
    if file_exists is None:
        file_exists = _exists

    entries = []
    pkg_path = os.path.join(cwd, "package.json")
    if file_exists(pkg_path):
        try:
            entries.extend(parse_package_json(read_file(pkg_path)))
        except Exception:
            pass
    py_path = os.path.join(cwd, "pyproject.toml")
    if file_exists(py_path):
        try:
            entries.extend(parse_pyproject(read_file(py_path)))
        except Exception:
            pass

    if not entries:
        return None

    shown = entries[:MAX_ENTRIES]
    more = f" (+{len(entries) - MAX_ENTRIES} more)" if len(entries) > MAX_ENTRIES else ""
    return (
        "## Installed dependency versions\n"
        "Use these exact versions — do not assume newer/older APIs:\n"
        + "\n".join(f"- {e}" for e in shown)
        + more
        + "\n"
    )


if __name__ == "__main__":
    result = run()
    if result:
        sys.stdout.write(result)
